### PR TITLE
Improvements for next mandatory

### PR DESCRIPTION
**Pull request by tomasbrod**
Status: **testing**.
Attempt 2 of #689. Improvements for smoother mandatory upgrade.
Add V9 block foundation.
Tie forking changes to block version 9: #524 (Beaconchecks) and #577 (ClearCache).
Improve storage of contract TrxID for debug purposes.
Fix newbie stake reward #552, @Foggyx420 PR was outdated.
Finally prevent duplicate superblocks for @Erkan-Yilmaz.

Note: a tally fix, which was included in #689, is *not included* in this PR. IMO it should be released in next update too.

Hope it works.

### CODE REVIEW

**Comment by tomasbrod** [comment]
Updated to reflect recent changes.

File: src/main.cpp [hunk 1]
```diff
@@ -8255,7 +8278,8 @@ bool MemorizeMessage(const CTransaction &tx, double dAmount, std::string sRecipi
                             fMessageLoaded = true;
                         }
 
-                        WriteCache(sMessageType,sMessageKey+";TrxID",tx.GetHash().GetHex(),nTime);
+                        if(fDebug)
+                            WriteCache("TrxID;"+sMessageType,sMessageKey,tx.GetHash().GetHex(),nTime);
```
Comment (iFoggz) [inline src/main.cpp:8282]: i like this i was thinking of the same. putting that in front.

**Comment by Mercosity** [comment]
Looks like we have progress .. Any idea when release will happen?

**Comment by tomasbrod** [comment]
@Mercosity Only after testing is successful. Testing is currently blocked by #668 Tally improvements unintentional changes breaking pre-mandatory syncing.

**Comment by Mercosity** [comment]
Ok @tomasbrod .. Thx for that .. waiting with anticipation ..

**Pull request merged**